Implement Token Expiry Check and Refresh using Refresh Tokens

krezovic opened the pull request:
<!-- Please describe your pull request here. -->

This merge request adds support for checking expiration of issued access tokens and  token refresh using refresh tokens

Strict Token Expiration will result in a 401 when access token has been expired and either no refresh token has been provided or refresh token itself has expired or has been revoked. This option is not enabled by default. I am not sure if the user should be automatically logged out in this case and I'm even more unsure on how to achieve this via code besides HTTP 302 to /logout or whatever the endpoint is.

Access Token refresh is automatically enabled when using well known configuration and the server supports "refresh_token" grant. Otherwise, it has to be manually enabled via checkbox toggle. In addition to refreshing the access token, the id token and user info is also refreshed and any new or updated groups are added for the currently logged in user.

Fixes: #100 

### Testing done

<!-- Comment:
Provide a clear description of how this change was tested.
At minimum this should include proof that a computer has executed the changed lines.
Ideally this should include an automated test or an explanation as to why this change has no tests.
Note that automated test coverage is less than complete, so a successful PR build does not necessarily imply that a computer has executed the changed lines.
If automated test coverage does not exist for the lines you are changing, you must describe the scenario(s) in which you manually tested the change.
For frontend changes, include screenshots of the relevant page(s) before and after the change.
For refactoring and code cleanup changes, exercise the code before and after the change and verify the behavior remains the same.
-->

```[tasklist]
### Submitter checklist
- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Link to relevant pull requests, esp. upstream and downstream changes
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue
```

<!--
Put an `x` into the [ ] to show you have filled the information.
The template comes from https://github.com/jenkinsci/.github/blob/master/.github/pull_request_template.md 
You can override it by creating .github/pull_request_template.md in your own repository 
-->

krezovic commented:
Hi @michael-doubez, this is a working draft where the most basic and common functionality works:

- Added "Enable/disable refresh token" support - a simple checkbox when "manual config", detected when "automatic config" by querying "supproted_grant_types" discovery response attribute
- Added "strict" mode (disabled by default) that indicates whether an expired access token and missing refresh token, or simply failed refresh request (due to various reasons) will result in "401" or we simply ignore it
- Added configurable clock skew that prolongs lifetime of an access token for a pre-configured amount of seconds to adjust for any time sync issues. Defaults to 60 seconds
- Any updated roles/groups are reflected on every token refresh, as they should be

Now, before I move on to update documentation and write some tests, I'd like to get some feedback if the approach is correct. In addition, I have additional points that I have noticed but not sure if I should handle or how to handle at all.

- First and most obvious point: In the build pipeline there is a warning that "accessToken" and "refreshToken" look sensitive and may be stored in plain text files. I have no idea where the SecurityContext is stored, so do you think we should mark the tokens as credentials?
- There is a concurrency issue where two parallel requests will trigger two token refresh requests at the same time. I am unable to find a good place for a synchronization object that does not involve a manual clean up. Should I take care about this? If yes, do you happen to know if there's any place that I could place a synchronization primitive such as timestamp of first request that triggered the refresh, and to let any subsequent parallel requests just skip the refresh as it's "being handled"? Is there any "self-cleaning" cache used elsewhere in the project that I could hook into?

Just for fun, without any activity I keep my jenkins tab opened and UI triggers a lot of requests every x seconds, resulting in at least 2 token refresh requests without any activity

![image](https://github.com/jenkinsci/oic-auth-plugin/assets/19533104/ea163a33-42da-4e5f-98ab-b848464f5147)


- When a Token is expired, no refresh can be performed, and strict mode is on, should we simply redirect the user to the logout URL or do we clean up the session manually and simply "halt"?

Any additional feedback is welcome.

michael-doubez commented:
Hello,

Thanks for the work. You can ignore the warnings about secret leaking, they are mostly irrelevant.
I'll check the concurrency issue but if the information is local to a user context, there shouldn't be any issue.

There are a lot of features in the PR. I would break them down into:

- [ ] handle session timeout: if refresh/offline token is unavailable or expired, how to clear session information and request re-login
- [ ] handle additional (refresh) token lifetime : how to request it if available, as you did. It should be store in session information such as to be available globally, it may also be revoked at logout because lifetime is typically longer than access token
- [ ] handle refresh of user info

I should be able to review in depth and (hopefuly) provide relevant feedback next Sunday.

krezovic commented:
Hello,

I believe I have handled most of the use-cases except the logout. The question remains on how do we proceed.

Answering the points in the order they were raised

- Easiest way to do this is to request log out at IdP. This is an already supported feature. This will invalidate the refresh token, and next refresh will fail. This use-case has been handled and works without clearing the Jenkins related session stuff (as of yet - we have to agree on how to proceed)
- Refresh tokens may be extended indefinitely as long as the session is active on the IdP. I wouldn't go beyond what was done here as described in step 1 - refresh token returns "invalid_grant" -> assume RT is expired or revoked -> destroy the user session. User may be logged elsewhere out of IdP and detecting this is a bit more bothersome than I'd like to go into - but there are standards on how to handle this scenario (see below)
- Refresh of user info and ID token has already been incorporated in the MR and I re-used the existing userinfo request code and IdToken decoding (without the validations - but that can be re-used if needed, as well)

Back to the first point, we can implement additional features if needed, namely

- We implement Token Revocation on manual logout when redirect to IdP has been disabled https://datatracker.ietf.org/doc/html/rfc7009
- We add support for OpenID Front and/or Back Channel Logout https://openid.net/specs/openid-connect-frontchannel-1_0.html / https://openid.net/specs/openid-connect-backchannel-1_0.html
- We add support OpenID Session Management and check_session_iframe, although I'm not sure how widely it's supported. Not even latest Spring Security supports it https://openid.net/specs/openid-connect-session-1_0.html

michael-doubez commented:
@krezovic the change looks good to me.

What bothers me most is the idtoken kept in the session information. It should be removed and kept only in the credentials but that can be ironed out later.

krezovic commented on src/main/java/org/jenkinsci/plugins/oic/OicCredentials.java:
```diff
@@ -0,0 +1,62 @@
+import hudson.Util;
+import java.time.Clock;
+
+public class OicCredentials {
```
Good catch. Thanks for the hint.

krezovic commented:
Thanks for the feedback @michael-doubez. I have removed IdToken from Session and moved it strictly to SecurityContext.

Thank you @jglick for spotting missing Serializable and for bumping this MR as I completely forgot about it.

With that said and done, the MR is now ready for review and hopefully merging.

jglick commented on src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java:
```diff
@@ -764,6 +817,21 @@ public String getAuthenticationGatewayUrl() {
+            public void doFilter(ServletRequest request, ServletResponse response, FilterChain chain)
+                    throws IOException, ServletException {
+
+                if (OicSecurityRealm.this.handleTokenExpiration(
+                        (HttpServletRequest) request, (HttpServletResponse) response)) {
+                    chain.doFilter(request, response);
```
It is not obvious (to me) from code inspection that this effectively runs before `BasicHeaderApiTokenAuthenticator`. The expected behavior, which I think could be asserted in `PluginTest` without new machinery beyond `TEST_USER_GROUPS_REFRESHED` etc., is that if a user’s groups change in the IdP and the previous token has expired, then a request authenticated using a Jenkins API token will in fact run under an `Authentication` reflecting that update to `authorities`. Or indeed if the user becomes completely invalid (e.g., removed from IdP realm) that a `UsernameNotFoundException` is thrown and the API token becomes unusable.

See also https://github.com/jenkinsci/oic-auth-plugin/pull/333#discussion_r1630141770: we no longer even define a custom `UserDetailsService` (what the API token processing code calls into), but on the other hand `handleTokenRefreshResponse` calls `loginAndSetUserData` which calls `fireLoggedIn` which probably accomplishes the goal. It is just not covered by the existing tests that I can see.

jglick replied:
From some preliminary experiments run by a colleague (@mikecirioli) it seems this does not in fact work, and could not work because the `handleTokenExpiration` method would be called in the context of an HTTP request with no servlet session (or a freshly generated empty session) and thus no `OicCredentials` and just return immediately without doing anything. In fact no approach based on keeping the refresh token (and whatever other information might be required for a refresh request) in a session attribute, or generally in memory, could possibly work; you would have to save everything required to disk, conventionally in a hidden `UserProperty`.

This PR purports to fix #100, which warns

> session cookies, API Tokens and so on may be valid forever, even if the user has long been disabled in the Identity Provider

It seems like what is fixed here is only the security of session cookies but not of API tokens; and #100 was originally filed as a security issue in response to #41, about handling API tokens.

Consider the scenario that a Jenkins user while part of some team creates an API token, perhaps for legitimate purposes, and saves it on a personal laptop. A month later they are fired and removed from the identity provider, or remain in the system but are reassigned to an unrelated team and the corresponding “group” removed from the IdP. A month after that (assume the controller has been restarted several times), _without attempting to log in to the controller again_, they check their API token (`curl`, `jenkins-cli.jar`) and find that it still works and still grants them the same permissions to reconfigure jobs or whatever they had while on the team. The only way to revoke access is for the Jenkins administrator to delete their user record from Jenkins or otherwise explicitly expire their API tokens.

krezovic replied:
Hi, I'm still processing what you asked - but I can assure you the filter is executed correctly and right after the default security handling defined in SecurityRealm::createFilter. You may notice that the new filter is plugged right after the default filters, which also includes loading of security context from session, distributed storage or wherever it's stored. I'd like that you share how the mentioned colleague deduced that the token refresh would not happen. I do note that you mention session a lot - however I only work with Spring SecurityContext which may or may not be stored in the session.

![image](https://github.com/jenkinsci/oic-auth-plugin/assets/19533104/5b99b75b-15bb-43e9-aa04-d0e4b02bc7e8)

![image](https://github.com/jenkinsci/oic-auth-plugin/assets/19533104/953f6123-1327-4540-b7e5-fe053a62f693)

krezovic replied:
I only have Keycloak that I can use for testing, and if I terminate my sessions via Keycloak console, I will get the "user refresh has expired"

Before:

![image](https://github.com/jenkinsci/oic-auth-plugin/assets/19533104/998e5327-ea03-4db0-baf5-efb52c1a27b2)

After:

![image](https://github.com/jenkinsci/oic-auth-plugin/assets/19533104/14e9e141-74e5-4aed-9e82-c072ab41708c)

Which yields the expected result in Jenkins:

![image](https://github.com/jenkinsci/oic-auth-plugin/assets/19533104/22382556-4666-4c07-8258-009cf1049dbd)

![image](https://github.com/jenkinsci/oic-auth-plugin/assets/19533104/9f082ab5-934f-4bdf-80e5-fab2ec347ba1)

krezovic replied:
> It is not obvious (to me) from code inspection that this effectively runs before `BasicHeaderApiTokenAuthenticator`.

I have just verified this - the method runs _after_ BasicHeaderApiTokenAuthenticator, which is defined in the parent createFilters method. Should it run _before_?

jglick replied:
I am not sure what the last screenshot is about. The point is to use an existing API token after a user is removed from Keycloak, without having accessed the controller using a browser in the interim.

> I only work with Spring SecurityContext which may or may not be stored in the session.

Right, but so far as I can tell after the initial login flow this context would be stored in the session by `HttpSessionContextIntegrationFilter2` / `HttpSessionSecurityContextRepository` for requests coming from a web browser with a session cookie. A REST or CLI request would bypass that, so the next filter to run would be `BasicHeaderProcessor` calling `BasicHeaderApiTokenAuthenticator`. That would ask the `UserDetailsService` for an `Authentication`, which with this plugin would be the default `UserDetailsServiceProxy` with no `delegate`, always throwing `UserMayOrMayNotExistException2`; `getUserDetailsForImpersonation2` would use `ImpersonatingUserDetailsService2` which catches that and falls back to `LastGrantedAuthoritiesProperty`, picking up the groups from the last web login. That would result in a `UsernamePasswordAuthenticationToken` with a `String` principal, not a `OicCredentials` as needed to trigger the refresh flow.

Unconfirmed so far, we will try to demonstrate the issue with a test.

krezovic replied:
No need to demonstrate, I have figured out what you talked about and will try to implement it to support the mentioned use-case.

jglick replied:
> if the token and/or user is no longer valid, then a request with such API Key must return 401

Right; and along the same lines, if the user is still valid but now has different group claims (at least after some expiration time has elapsed), the authentication in effect when using the API token should reflect the updated `authorities` (since these might have been used by the `AuthorizationStrategy` to make access control decisions).

krezovic replied:
I believe I may have ran into a problem with user initiated logout. The process will invalidate the AT/RT and even if it's saved, the API key call will run into a 401 in this case. How to handle this scenario?

jglick replied:
Not sure if there is a solution, other than to tell users who need continued access to API tokens to not log out.

krezovic commented:
Thanks to @jglick  for brief explanation on what I've missed and the original purpose of issue #100.

The plugin will now serialize OIDC tokens into User object so they are valid when accessing the instance with API Keys.

Logging with an API Key will also trigger token refresh if AT has expired. If RT has also expired or has been revoked, the  request made with API Key will correctly return 401.

Downside of this is that all API Keys will automatically become invalid when "Log out from OIDC Provider" option has been selected. For this scenario, I have nullified all token values in this scenario so no token refresh will be attempted but the expiration logic will correctly detect an expired token.

mikecirioli commented:
I've been playing with this in a local cluster and the easiest way i've found is to:

- Create a job or folder that your test user should be able to access
- set the AT expiration to something short (60s) 
- use the cli command `list-jobs` (using an access token) - verify that you can see the job
- wait 60 seconds (or manually invalidate the access token
- use the cli command `list-jobs` (using an access token) - verify that you now get a 403 type response
You can also use the HTTP api to do the same `list-jobs` check

michael-doubez reviewed (approved):
LGTM

krezovic commented on src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java:
```diff
@@ -1284,6 +1384,107 @@ public HttpResponse doFinishLogin(StaplerRequest request) throws IOException {
+            if (isUseRefreshTokens() && credentials.getRefreshToken() != null) {
+                refreshExpiredToken(credentials, httpServletRequest);
+            } else if (isStrictTokenExpirationCheck()) {
+                throw HttpResponses.errorWithoutStack(401, "Token expired");
```
To avoid repeating myself, all the comments reference outdated code except this one.

Sadly, I'm conflicted what to do when logging in via Browser or via an API Key.

krezovic commented on src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java:
```diff
@@ -1284,6 +1384,107 @@ public HttpResponse doFinishLogin(StaplerRequest request) throws IOException {
+            handleTokenRefreshResponse(flow, credentials, tokenResponse);
+
+            if (oicSession != null && tokenResponse.getIdToken() != null) {
+                oicSession.setIdToken(tokenResponse.getIdToken());
```
Already moved.

krezovic commented on src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java:
```diff
@@ -1284,6 +1384,107 @@ public HttpResponse doFinishLogin(StaplerRequest request) throws IOException {
+            userInfo = getUserInfo(flow, tokenResponse.getAccessToken());
+        }
+
+        String username = determineStringField(userNameFieldExpr, idToken, userInfo);
```
Already done

krezovic commented on src/main/java/org/jenkinsci/plugins/oic/OicCredentials.java:
```diff
@@ -0,0 +1,62 @@
+        this.accessToken = accessToken;
+        this.idToken = idToken;
+        this.refreshToken = refreshToken;
+        this.clock = Util.fixNull(clock, Clock.systemUTC());
```
Already done.

krezovic commented:
> I've been playing with this in a local cluster and the easiest way i've found is to:
> 
> * Create a job or folder that your test user should be able to access
> * set the AT expiration to something short (60s)
> * use the cli command `list-jobs` (using an access token) - verify that you can see the job
> * wait 60 seconds (or manually invalidate the access token
> * use the cli command `list-jobs` (using an access token) - verify that you now get a 403 type response
>   You can also use the HTTP api to do the same `list-jobs` check

Hi. I was more asking in context of "integration test" jenkins instance (in PluginTest, annotated with `@JenkinsRule`). Which is why I said I'd prefer a REST API and that I do not have to configure or enable security manually on the test instance - not sure what the consequences will be.

Do note that I can easily call the HTML page users/testUser (which is the uid of the test user that's created), but this works both with authentication and without any authentication - so I'm kinda in conflict.

mikecirioli commented:
I'll try to take a look later today or this weekend and see if i can help contribute a test.  So far, i have not been able to successfully get the plugin to refresh my users authorities after i have removed them from the IDP, and access token remains working.

Strict token expiration does successfully revoke jenkins api token access when the primary token expires.

michael-doubez commented on src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java:
```diff
@@ -1252,6 +1346,155 @@ public HttpResponse doFinishLogin(StaplerRequest request) throws IOException {
+        IdToken parsedIdToken;
+
+        try {
+            parsedIdToken = tokenResponse.parseIdToken();
```
The `failedCheckOfTokenField()` should still be enforced.

mikecirioli commented on src/main/resources/org/jenkinsci/plugins/oic/OicSecurityRealm/help-logoutFromOpenidProvider.html:
```diff
@@ -1,3 +1,3 @@
 <div>
-    Logout from Jenkins entails logout from OpenID Provider.
+    Logout from Jenkins entails logout from OpenID Provider. Please note that this will make all API Keys invalid.
```
I understand why this is, but it may be problematic for some users who would expect that their API token should still function as a valid means of access even if they have logged out from the OpenID provider

michael-doubez replied:
This not the intent ; the logout here applies when the user logout from Jenkins.
User cannot reasonably assume the token should still work.

It is rather the other way around, when the user logs out of Jenkins, we make the logout from OP optional because:

> https://openid.net/specs/openid-connect-rpinitiated-1_0.html#RPLogout
> At the Logout Endpoint, the OP SHOULD ask the End-User whether to log out of the OP as well. [...]

But some OP logs out the user from of all RPs.

michael-doubez replied:
Wait. Are we talking about jenkins API token ?

mikecirioli replied:
yes, jenkins api tokens - sorry for the confusion

krezovic replied:
This is a necessary stalemate. See https://github.com/jenkinsci/oic-auth-plugin/pull/310#discussion_r1638780197 for details and reasoning. If you have something else in mind, please do let me know. cc: @jglick

mikecirioli replied:
nope @krezovic , i have come to the same conclusion (thanks for your patience!)

mikecirioli commented on src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java:
```diff
@@ -1186,9 +1263,26 @@ public void doLogout(StaplerRequest req, StaplerResponse rsp) throws IOException
+            if (this.logoutFromOpenidProvider && !Strings.isNullOrEmpty(this.endSessionEndpoint)) {
+                // This ensures that token will be expired at the right time with API Key calls, but no refresh can be
+                // made.
+                user.addProperty(new OicCredentials(null, null, null, credentials.getExpiresAtMillis()));
```
Invalidating the refresh token here seems to cause jenkins to fall back to `lastGrantedAuthorities` for the user when doing token based access.  This means (afaict in my testing) that tokens will still be effectively valid even if  that access was removed at the idp level.  The new authorities are only propagated to jenkins if the user (whose access was removed) attempts to log back in again.

If the `refreshToken` is left persisted in the user properties then it will be used to refresh the API token (once the id token as expired) and at that point it will get the updated authorities for the user associated with the token.  For example, if their access to the particular jenkins instance was updated at the IdP then this will correctly get reflected in their jenkins user's `grantedAuthorities`

mikecirioli replied:
hmm, it seems that if the refresh token is saved then the next time the user tried to log directly into the instance then that token will be used and the credentials refreshed and the user will not be prompted to log back in at the idp, which doesn't really seem ideal either.

krezovic replied:
> Invalidating the refresh token here seems to cause jenkins to fall back to `lastGrantedAuthorities` for the user when doing token based access. This means (afaict in my testing) that tokens will still be effectively valid even if that access was removed at the idp level. The new authorities are only propagated to jenkins if the user (whose access was removed) attempts to log back in again.
> 
> If the `refreshToken` is left persisted in the user properties then it will be used to refresh the API token (once the id token as expired) and at that point it will get the updated authorities for the user associated with the token. For example, if their access to the particular jenkins instance was updated at the IdP then this will correctly get reflected in their jenkins user's `grantedAuthorities`

This is only true if "strict expiration check" is disabled - which is default at the moment. Once it's changed to require refresh token by default, user will be greeted with 403.

krezovic replied:
> hmm, it seems that if the refresh token is saved then the next time the user tried to log directly into the instance then that token will be used and the credentials refreshed and the user will not be prompted to log back in at the idp, which doesn't really seem ideal either.

I'm not sure why do you see this a problem. This is exactly the purpose of refresh token. However, if the user logs out at IdP, the refresh token will be invalidated and the user will be greeted with 403/redirect to login.

michael-doubez commented on src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java:
```diff
@@ -195,6 +209,14 @@ public static enum TokenAuthMethod {
+
+    /** Flag to enable strict token expiration check
+     */
+    private boolean strictTokenExpirationCheck = false;
```
On second thought, I'd pefer this option to by true by default or even renamed to disableTokenExpirationCheck.

IMHO it is better to enforce security by default and allow user to bypass it if they want (unless it is added security or it breaks usability too much - like PKCE).

michael-doubez commented on src/main/java/org/jenkinsci/plugins/oic/OicSecurityRealm.java:
```diff
@@ -1186,9 +1263,26 @@ public void doLogout(StaplerRequest req, StaplerResponse rsp) throws IOException
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+        User user = User.get2(authentication);
+
+        Assert.notNull(user, "User must not be null");
```
A null user at logout is not a death sentence - let just ignore that and assume the user is already logged out.

A notice a null id token is not handleld in getPostLogOutUrl2().

krezovic replied:
For "null user", it's to silence spotbugs - this use case is unlikely as it fetches the user from user storage. It is kinda "end of the world" if the user we're trying to log out is not present in the storage.

```
[INFO] --- spotbugs:4.8.3.1:check (spotbugs) @ oic-auth ---
[INFO] BugInstance size is 1
[INFO] Error size is 0
[INFO] Total bugs: 1
[ERROR] Medium: Possible null pointer dereference in org.jenkinsci.plugins.oic.OicSecurityRealm.doLogout(StaplerRequest, StaplerResponse) due to return value of called method [org.jenkinsci.plugins.oic.OicSecurityRealm, org.jenkinsci.plugins.oic.OicSecurityRealm] Dereferenced at OicSecurityRealm.java:[line 1271]Known null at OicSecurityRealm.java:[line 1270] NP_NULL_ON_SOME_PATH_FROM_RETURN_VALUE
```

For ID Token - check has been added and tests have been updated.

michael-doubez reviewed (approved):
LGTM

mikecirioli reviewed (approved):
Looks good to me, thanks for the discussion

mikecirioli commented:
@krezovic what else needs to be done in order to merge this PR?  Is there something you need a hand with or are you just waiting for me feedback?

michael-doubez commented:
@krezovic conflicts needs to be resolved.

IMHO we can merge it as it is.

michael-doubez commented:
@krezovic I have been working on the merge, I should be able to contribute it before the end of the week

michael-doubez commented:
See https://github.com/michael-doubez/oic-auth-plugin/tree/token_refresh

krezovic commented:
> See https://github.com/michael-doubez/oic-auth-plugin/tree/token_refresh
> 
> Note: some unit test still to adapt/fix

Hi. Thanks for doing the heavy lifting. I was already halfway through and was hoping to finish it over the weekend. Your changes to test really helped and now I have finished fixing the tests and took a bit different approach than you in passing the nullable refresh token.

michael-doubez reviewed (commented):
LGTM
